augur/merge.py

```diff
@@ -503,3 +579,85 @@ def quote(s):
     # …and instead quote a final empty string down here if we're still empty
     # after joining all our parts together.
     return quoted if quoted else shquote('')
+
+
+def run_command(command: str, print_stdout: bool):
+    """Run a command, clearly marking any messages that resemble SeqKit output."""
+    process = subprocess.Popen(
+        ['bash', '-euo', 'pipefail', '-c', command],
+        stdout=subprocess.PIPE if print_stdout else subprocess.DEVNULL,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+
+    if print_stdout:
+        for line in iter(process.stdout.readline, ''):
+            print(line.rstrip())
+
+    for line in iter(process.stderr.readline, ''):
+        # Remove ANSI escape sequences common in SeqKit output.
+        line = re.sub(r'\x1b\[[\d;]+m', '', line)
+
+        # Detect messages from seqkit and append a prefix.
+        if line.startswith("[INFO]"):
+            print_debug(f"[SeqKit] {line.rstrip()}")
+        elif line.startswith("[ERRO]"):
+            print_info(f"[SeqKit] {line.rstrip()}")
+        else:
+            print_info(line.rstrip())
+
+    # Wait for the seqkit process to finish.
+    # FIXME: address potential deadlock issue?
+    # <https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait>
+    # I think this is safe since process.stdout and process.stderr are iterated manually.
+    return process.wait()
+
+
+def seqkit():
+    """
+    Internal helper for invoking ``seqkit``, the SeqKit CLI program.
+
+    Unlike ``sqlite3()``, this function is not a wrapper around subprocess.run.
+    It is meant to be called without any parameters and only returns the
+    location of the executable. This is due to differences in the way the two
+    programs are invoked.
+    """
+    seqkit = shquote(os.environ.get("SEQKIT", which("seqkit")))
+    if not seqkit:
+        raise AugurError(dedent(f"""\
+            Unable to find the program `seqkit`.  Is it installed?
+            In order to use `augur merge`, the SeqKit CLI must be installed
+            separately.  It is typically provided by a Nextstrain runtime.
+            """))
+    return seqkit
+
+
+def parse_seqkit_dup_num_file(contents: str):
+    """Extract duplicated IDs from the output of seqkit rmdup --dup-num-file.
+
+    There is one line per duplicated ID, in the format of a number n followed by
+    tab character followed by the ID duplicated n times, separated by ", ".
+
+    Example FASTA headers
+
+        > id, 1
+        > id, 2
+        > id, 3
+        > id1
+        > id1
+
+    will produce a --dup-num-file output file
+
+        3\tid,, id,, id,
+        2\tid1, id1
+
+    which parsed by this function will yield
+
+        id,
+        id1
+
```

Comment:
_non-blocking_

This sort of function is the perfect candidate for doctests, IMO, so they stay close to the code.

Reply:
Doctests added:

https://github.com/nextstrain/augur/blob/e34a62bdc43a1e12cd8b58277f750a27bb2e20e7/augur/merge.py#L631-L649